### ISSUE TITLE
Unknown directive

### ISSUE THREAD

**Issue by PyB1l** [LilSpazJoekp/docstrfmt#45]
Is there a way to skip unknown directives?  I am using mermaid plugin, and I am getting the following error

``` 
Unknown directive type "mermaid".
```

**Comment by LilSpazJoekp** [comment]
Seems reasonable. I'll see if I can skip them or dynamically add them. Would you mind providing an example to reproduce this?

**Comment by PyB1l** [comment]
This is an example mermaid plugin directive

```

.. mermaid::
   :caption: Figure 1.1 - Computation Algorithm flow

   flowchart LR
        params((User Orders)) --> WF[WorkFlow Manager <br> Library]

```

**Comment by PyB1l** [comment]
There several sphinx plugins with different custom directives . I suppose a valid roadmap would be to ignore
unknown directives until you can completely support a specific sphinx plugin

**Comment by LilSpazJoekp** [comment]
Yeah either that or I'd like to add a way to programmatically detect plugins and add those at the start.

**Comment by aarnphm** [comment]
Hi, we at https://github.com/bentoml/BentoML are looking into RST formatter and really like your project. We would love to have this feature supported and would love to help out if needed.

**Comment by LilSpazJoekp** [comment]
Help would be greatly appreciated! The best solution for arbitrary extensions/directives is to hook into Sphinx and load in the directives directly. The next best is to allow the use to tell docstrfmt to ignore unknown ones. I plan on taking a look tomorrow and see if there are any directives that are missing that are in your project and see if they can be added in easily.

**Comment by LilSpazJoekp** [comment]
@aarnphm Just briefly looking at your project, I see that you are using napoleon formatted docs. This won't work without some work but I think it possible.

**Comment by aarnphm** [comment]
sounds good, let me know if you need any help 😄

**Comment by PyB1l** [comment]
Do we have any updates on this?

**Comment by LilSpazJoekp** [comment]
I don't unfortunately. Been busy recently, sorry about that. 

The directives you're using can be added [here](https://github.com/LilSpazJoekp/docstrfmt/blob/master/docstrfmt/rst_extras.py#L115). If you'd like to open a PR I'll merge it in really quick.

**Comment by PyB1l** [comment]
Is there a way to skip a custom directive from formating?

**Comment by LilSpazJoekp** [comment]
Unfortunately it's not feasible to just skip the unknown directives. The formatter doesn't have the ability to just pass through a node untouched.

What I can do is add the directive really quick to support mermaid blocks. Is `mermaid` the only directive you need?

**Comment by aarnphm** [comment]
Right now, I have a hack around adding support for additional directives. You can see it here https://github.com/LilSpazJoekp/docstrfmt/compare/master...aarnphm:docstrfmt:feat/ext.

But I think there might be something we can do here to support external directives.

Maybe we can have a plugin design for it.

**Comment by PyB1l** [comment]
I  cloned the repo and I am testing the plugin implementation.
In the meantime I noticed that `csv-table`  directive is translated to table

**Comment by FFY00** [comment]
I have run into the same issue with the `caption` directive for `code-block`  in sphinx.

```
$ docstrfmt docs/new/tutorials/introduction.rst
ERROR: File "/home/anubis/git/mesonpy/docs/new/tutorials/introduction.rst":
Error in "code-block" directive:
unknown option: "caption".
Failed to format '/home/anubis/git/mesonpy/docs/new/tutorials/introduction.rst'
1 file were checked.
Done, but 1 error occurred ❌💥❌
```

> Unfortunately it's not feasible to just skip the unknown directives.

That's unfortunate, as it means we'll always be playing whack-a-mole to keep up with new directives.
`Should I open a new issue for this directive?

**Comment by LilSpazJoekp** [comment]
@FFY00 
> Should I open a new issue for this directive?

Yeah

> That's unfortunate, as it means we'll always be playing whack-a-mole to keep up with new directives.

It is. I'd exploring an option to dynamically load in from sphinx by partially initializing sphinx. During initialization it will load in the settings of the project and any extensions it is using.

**Comment by LilSpazJoekp** [comment]
@aarnphm 
> Maybe we can have a plugin design for it.

I like this idea. Let me take a look at your branch and see if it's something that could work.

**Comment by LilSpazJoekp** [comment]
@PyB1l 
> Yes, I cloned the repo and registered the mermaid plugin myself.
> formatting raises no error, but it messes with text under the directive.

Did you specify `raw=True`?

**Comment by FFY00** [comment]
@LilSpazJoekp actually, I was looking at https://github.com/dzhu/rstfmt/pull/19, and it seems viable since sphinx marks the bad nodes and allows you to access the raw source. 

With this approach, we'd probably want to give a warning, as the unknown directives are kept as-is instead of being formatted, but this seems better than just blowing up like it happening right now. What do you think?

**Comment by LilSpazJoekp** [comment]
Interesting, I wasn't aware of that. Yes, that would be viable and have it raise a warning instructing the user to open an issue for formatted support.

**Comment by aarnphm** [comment]
I think we can do something about a ABC implementation for sphinx plugins. It would be very neat.

Right now on my branch it is a hack solution that I just run a container to fix all of our docs format.

**Comment by LilSpazJoekp** [comment]
I have an almost working dynamic sphinx loader: #52 if you guys want to take a look. There's some bugs with it but it might be a viable solution. In its current state, it can drop parts out of the docs.

**Comment by LilSpazJoekp** [comment]
> We can do something about a ABC implementation for sphinx plugins. It would be very neat.

I really like this idea, however I'm not sure the best way to implement something like that. The way the formatter works is it looks up the type of the node it's formatting from the Formatters class. We could provide a way to dynamically add new functions to that. Basically we wouldn't have to add a new method for every custom directive/role type as it's still a directive/role at the end of the day.

**Comment by aarnphm** [comment]
We can also think about a dispatcher for new custom directive/role for python API and do some codegen 😄 

```python
import docstrfmt

from docstrfmt.gen import make_node_lookup_function

formatter = docstrfmt.Formatter()

formatter.add_directives(MyDirectiveClass, lookup_function=make_node_lookup_function(...)
```

food for thought

**Comment by LilSpazJoekp** [comment]
That could work. I don't think each custom type would need to have it's own formatting function since I don't think you can make your own constructs. Like you can have custom directives and roles but they still have to follow the rules of those constructs at the end of the day.

Basically, I'd like to find a way to dynamically load in the sphinx extensions they have in their conf.py file and register each extension's directives and roles. They will either have formattable content (e.g., `.. note::`, `.. code-block::`) or not (e.g., `.. raw::` or `.. mermaid::`). Currently, you can have the custom types have their own function–note, warning, and danger directives technically do have their own but they all execute the same formatting method.

**Comment by FFY00** [comment]
Any updates?

**Comment by julian-poidevin** [comment]
Would love to see this implemented. [`rstfmt` as the same limitation ](https://github.com/dzhu/rstfmt/issues/14).

**Comment by mpaxson** [comment]
any updates?

**Comment by LilSpazJoekp** [comment]
Unfortunately no. If there is a specific directive needed I am more than happy to add support for it. However, for the time being, I do not have the bandwidth to do a full overhaul of how directives from arbitrary sphinx contrib modules are resolved. I did start #52 to do this but I wasn't able to get it fully working.

**Comment by johschmitz** [comment]
Directives I found preventing a usage of this tool:

- todo
- uml
- asciinema
- index
- glossary

It would be totally fine to ignore them and continue.

I also found some unknown options
- linenos
- lineno-start
- emphasize-lines
- caption

And an unknown node
- rubric

There's also the problem to support things like Sphinx needs which come with a bunch of custom directives and also fully custom internal directives. It would be nice if there would be some option to ignore everything unknown, throwing a warning and just continuing.

**Comment by LilSpazJoekp** [comment]
I can look into adding those directives and role. What directive is the options for? `code-block`?

> There's also the problem to support things like Sphinx needs which come with a bunch of custom directives and also fully custom internal directives. It would be nice if there would be some option to ignore everything unknown, throwing a warning and just continuing.

The issue with just skipping unknown directives is it would result in the block being removed from the document entirely. The way the formatting works is to regenerate the entire document. Some processing has to be done to it to skip it. This is why dynamic support for sphinx (as in loading in sphinx and their configured plugins) is required. This is something I would like some help with. I have a partially functional solution in #52 but it needs some work still.

**Comment by johschmitz** [comment]
Yes, these are `.. code-block::` options.

**Comment by johschmitz** [comment]
Regarding the dynamic support and unknown things in general, I wonder how hard it would be to implement a "workaround" option that just copies everything unknown to the output as is and continues. Is this a hard parser problem due to the structure of the language?

**Comment by LilSpazJoekp** [comment]
> Yes, these are .. code-block:: options.

Support is planned for these.

**Comment by LilSpazJoekp** [comment]
In my experience with implementing that, yes. The parsing heavily
relies on docutils with sphinx support injected. The parsing happens first and if I recall correctly, it rejects unknown symbols and blocks as errors in a lossy manner.

> I wonder how hard it would be to implement a "workaround" option that just copies everything unknown to the output as is and continues.

If you're up for it, you're welcome to take a look and see if that's feasible.

**Comment by LilSpazJoekp** [comment]
Hello all! Just wanted to comment here that I was able to work out a solution for this issue. I've released it in version 1.8.0 and should be live on PyPI now!
